### PR TITLE
Add telemetry

### PR DESCRIPTION
**Pull request by mtenpow**
Sample output:

```
Testing against: playht-self-test-0000000000000012.on-prem.play.ht:11045
Acquiring lease
Acquired lease
format=wav, sample_rate=24.0kHz:
Start tts-request @ 1713380953.165771
Hey: StatusCode.RESOURCE_EXHAUSTED
Hey: StatusCode.RESOURCE_EXHAUSTED
Hey: StatusCode.RESOURCE_EXHAUSTED
Finish tts-request @ 1713380954.029474 duration: 0.8637030124664307
Sending text: This is a test. A very different test.
Start tts-request @ 1713380954.0296571
Hey: StatusCode.RESOURCE_EXHAUSTED
Hey: StatusCode.RESOURCE_EXHAUSTED
Hey: StatusCode.RESOURCE_EXHAUSTED
Received chunk 0 at t=213ms
Received chunk 1 at t=317ms <- first audio received
Received chunk 2 at t=318ms <- started playing audio
Received chunk 3 at t=584ms
Received chunk 4 at t=584ms
Received chunk 5 at t=584ms
Received chunk 6 at t=584ms
Received chunk 7 at t=584ms
Received chunk 8 at t=585ms
Received chunk 9 at t=585ms
Received chunk 10 at t=585ms
Received chunk 11 at t=588ms
Received chunk 12 at t=590ms
Received chunk 13 at t=590ms
Received chunk 14 at t=590ms
Received chunk 15 at t=680ms
Received chunk 16 at t=680ms
Received chunk 17 at t=681ms
Received chunk 18 at t=681ms
Received chunk 19 at t=765ms
Received chunk 20 at t=766ms
Received chunk 21 at t=766ms
Received chunk 22 at t=766ms
Received chunk 23 at t=851ms
Received chunk 24 at t=852ms
Received chunk 25 at t=853ms
Received chunk 26 at t=854ms
Received chunk 27 at t=943ms
Received chunk 28 at t=944ms
Received chunk 29 at t=945ms
Received chunk 30 at t=945ms
Received chunk 31 at t=1030ms
Received chunk 32 at t=1035ms
Received chunk 33 at t=1036ms
Received chunk 34 at t=1037ms
Received chunk 35 at t=1121ms
Received chunk 36 at t=1122ms
Received chunk 37 at t=1122ms
Received chunk 38 at t=1123ms
Received chunk 39 at t=1214ms
Received chunk 40 at t=1215ms
Received chunk 41 at t=1216ms
Received chunk 42 at t=1216ms
Received chunk 43 at t=1304ms
Received chunk 44 at t=1305ms
Received chunk 45 at t=1306ms
Received chunk 46 at t=1306ms
Received chunk 47 at t=1390ms
Received chunk 48 at t=1391ms
Received chunk 49 at t=1391ms
Received chunk 50 at t=1392ms
Received chunk 51 at t=1394ms
Received chunk 52 at t=1395ms
Finish tts-request @ 1713380955.4256299 duration: 1.395972728729248
format=wav, sample_rate=24.0kHz
Start tts-request @ 1713380959.8640919
Hey: StatusCode.RESOURCE_EXHAUSTED
Hey: StatusCode.RESOURCE_EXHAUSTED
Hey: StatusCode.RESOURCE_EXHAUSTED
Finish tts-request @ 1713380960.649908 duration: 0.7858161926269531
Sending text: This is a test. A very different test.
Start tts-request @ 1713380960.650038
Hey: StatusCode.RESOURCE_EXHAUSTED
Hey: StatusCode.RESOURCE_EXHAUSTED
Hey: StatusCode.RESOURCE_EXHAUSTED
Received chunk 0 (78 bytes) at t=189ms
Received chunk 1 (4096 bytes) at t=293ms <- first audio received
Received chunk 2 (4096 bytes) at t=295ms <- start playing audio
Received chunk 3 (4096 bytes) at t=399ms
Received chunk 4 (4096 bytes) at t=400ms
Received chunk 5 (4096 bytes) at t=400ms
Received chunk 6 (4096 bytes) at t=400ms
Received chunk 7 (4096 bytes) at t=480ms
Received chunk 8 (4096 bytes) at t=482ms
Received chunk 9 (4096 bytes) at t=482ms
Received chunk 10 (4096 bytes) at t=483ms
Received chunk 11 (4096 bytes) at t=568ms
Received chunk 12 (4096 bytes) at t=569ms
Received chunk 13 (4096 bytes) at t=569ms
Received chunk 14 (4096 bytes) at t=569ms
Received chunk 15 (4096 bytes) at t=657ms
Received chunk 16 (4096 bytes) at t=658ms
Received chunk 17 (4096 bytes) at t=658ms
Received chunk 18 (4096 bytes) at t=659ms
Received chunk 19 (4096 bytes) at t=742ms
Received chunk 20 (4096 bytes) at t=748ms
Received chunk 21 (4096 bytes) at t=749ms
Received chunk 22 (4096 bytes) at t=749ms
Received chunk 23 (4096 bytes) at t=829ms
Received chunk 24 (4096 bytes) at t=830ms
Received chunk 25 (4096 bytes) at t=831ms
Received chunk 26 (4096 bytes) at t=831ms
Received chunk 27 (4096 bytes) at t=920ms
Received chunk 28 (4096 bytes) at t=921ms
Received chunk 29 (4096 bytes) at t=921ms
Received chunk 30 (4096 bytes) at t=921ms
Received chunk 31 (4096 bytes) at t=1008ms
Received chunk 32 (4096 bytes) at t=1008ms
Received chunk 33 (4096 bytes) at t=1008ms
Received chunk 34 (4096 bytes) at t=1008ms
Received chunk 35 (4096 bytes) at t=1105ms
Received chunk 36 (4096 bytes) at t=1107ms
Received chunk 37 (4096 bytes) at t=1108ms
Received chunk 38 (4096 bytes) at t=1108ms
Received chunk 39 (4096 bytes) at t=1187ms
Received chunk 40 (4096 bytes) at t=1188ms
Received chunk 41 (4096 bytes) at t=1189ms
Received chunk 42 (4096 bytes) at t=1189ms
Received chunk 43 (0 bytes) at t=1189ms
Finish tts-request @ 1713380961.839787 duration: 1.189749002456665

{
  'operation': 'tts-request',
  'status': 'ok',
  'start_time': 1713380953.165771,
  'end_time': 1713380954.029474,
  'duration': 0.8637030124664307,
  'counters': {
    'retry': 2,
    'fallback': 1,
    'ok': 1
  },
  'attributes': {
    'text': [
      "['Warmup.']",
      "['Warmup.']",
      "['Warmup.']",
      "['Warmup.']"
    ],
    'endpoint': [
      'playht-self-test-0000000000000012.on-prem.play.ht:11045',
      'playht-self-test-0000000000000012.on-prem.play.ht:11045',
      'playht-self-test-0000000000000012.on-prem.play.ht:11045',
      'prod.turbo.play.ht:443'
    ],
    'retry.reason': [
      'StatusCode.RESOURCE_EXHAUSTED',
      'StatusCode.RESOURCE_EXHAUSTED'
    ],
    'fallback.reason': [
      'StatusCode.RESOURCE_EXHAUSTED'
    ]
  },
  'timers': {
    'time-to-first-audio': {
      'duration': 2.146191120147705,
      'count': 1
    },
    'retry-backoff': {
      'duration': 0.10383725166320801,
      'count': 2
    }
  }
}

{
  'operation': 'tts-request',
  'status': 'ok',
  'start_time': 1713380954.0296571,
  'end_time': 1713380955.4256299,
  'duration': 1.395972728729248,
  'counters': {
    'retry': 2,
    'fallback': 1,
    'ok': 1
  },
  'attributes': {
    'text': [
      "['This is a test. A very different test.']",
      "['This is a test. A very different test.']",
      "['This is a test. A very different test.']",
      "['This is a test. A very different test.']"
    ],
    'endpoint': [
      'playht-self-test-0000000000000012.on-prem.play.ht:11045',
      'playht-self-test-0000000000000012.on-prem.play.ht:11045',
      'playht-self-test-0000000000000012.on-prem.play.ht:11045',
      'prod.turbo.play.ht:443'
    ],
    'retry.reason': [
      'StatusCode.RESOURCE_EXHAUSTED',
      'StatusCode.RESOURCE_EXHAUSTED'
    ],
    'fallback.reason': [
      'StatusCode.RESOURCE_EXHAUSTED'
    ]
  },
  'timers': {
    'time-to-first-audio': {
      'duration': 0.9817550182342529
    },
    'retry-backoff': {
      'duration': 0.10982489585876465
    }
  }
}

{
  'operation': 'tts-request',
  'status': 'ok',
  'start_time': 1713380959.8640919,
  'end_time': 1713380960.649908,
  'duration': 0.7858161926269531,
  'counters': {
    'retry': 2,
    'fallback': 1,
    'ok': 1
  },
  'attributes': {
    'text': [
      "['Warmup.']",
      "['Warmup.']",
      "['Warmup.']",
      "['Warmup.']"
    ],
    'endpoint': [
      'playht-self-test-0000000000000012.on-prem.play.ht:11045',
      'playht-self-test-0000000000000012.on-prem.play.ht:11045',
      'playht-self-test-0000000000000012.on-prem.play.ht:11045',
      'prod.turbo.play.ht:443'
    ],
    'retry.reason': [
      'StatusCode.RESOURCE_EXHAUSTED',
      'StatusCode.RESOURCE_EXHAUSTED'
    ],
    'fallback.reason': [
      'StatusCode.RESOURCE_EXHAUSTED'
    ]
  },
  'timers': {
    'time-to-first-audio': {
      'duration': 1.0124499797821045
    },
    'retry-backoff': {
      'duration': 0.10796475410461426
    }
  }
}

{
  'operation': 'tts-request',
  'status': 'ok',
  'start_time': 1713380960.650038,
  'end_time': 1713380961.839787,
  'duration': 1.189749002456665,
  'counters': {
    'retry': 2,
    'fallback': 1,
    'ok': 1
  },
  'attributes': {
    'text': [
      "['This is a test. A very different test.']",
      "['This is a test. A very different test.']",
      "['This is a test. A very different test.']",
      "['This is a test. A very different test.']"
    ],
    'endpoint': [
      'playht-self-test-0000000000000012.on-prem.play.ht:11045',
      'playht-self-test-0000000000000012.on-prem.play.ht:11045',
      'playht-self-test-0000000000000012.on-prem.play.ht:11045',
      'prod.turbo.play.ht:443'
    ],
    'retry.reason': [
      'StatusCode.RESOURCE_EXHAUSTED',
      'StatusCode.RESOURCE_EXHAUSTED'
    ],
    'fallback.reason': [
      'StatusCode.RESOURCE_EXHAUSTED'
    ]
  },
  'timers': {
    'time-to-first-audio': {
      'duration': 0.8965945243835449
    },
    'retry-backoff': {
      'duration': 0.1061248779296875
    }
  }
}
```

### CODE REVIEW

File: pyht/telemetry.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        self.operation = operation
+        self.start_time = time.time()
+        print(f"Start {operation} @ {self.start_time}")
+        return self
+
+    def inc(self, counter: str, count: int = 1) -> Metrics:
+        if counter not in self.counters:
+            self.counters[counter] = 0
+        self.counters[counter] += count
+        return self
```
Comment (NCarrollPlay) [inline pyht/telemetry.py:75]: nit: prefer `.get()` or `.setdefault()` for dict access.
```python
def inc(self, counter: str, count: int = 1) -> 'Metrics':  # quote return type of self for backwards compat.
    self.counters[counter] = self.counters.get(counter, 0) + count
    return self
```

`.setdefault()` is for ref objects
```python
class Counter:
    def __init__(self, value: int = 0):
        self._v = value
    def __repr__(self):
        return repr(self._v)
    def __add__(self, v: int):
        self._v += v
        return self
    def __sub__(self, v: int):
        self._v -= v
        return self

def inc(self, counter: str, count: int = 1) -> 'Metrics':
    self.counters.setdefault(counter, Counter()) + 1
    return self
```

Comment (mtenpow) [reply pyht/telemetry.py]: Fixed.

File: pyht/async_client.py [hunk 2]
```diff
@@ -198,60 +201,83 @@ async def tts(
...
-
-                logging.info(f"Falling back to {self._fallback_rpc[0]} because {self._rpc[0]} threw: {error_code}")
+            for attempt in range(1, max_attempts + 1):
                 try:
-                    stub = api_pb2_grpc.TtsStub(self._fallback_rpc[1])
-                    stream: TtsUnaryStream = stub.Tts(request)
+                    metrics.append("text", str(request.params.text)).append("endpoint", str(self._rpc[0]))
+                    metrics.start_timer("time-to-first-audio", auto_finish=False)
+                    stub = api_pb2_grpc.TtsStub(self._rpc[1])
+                    chunks: TtsUnaryStream = stub.Tts(request)
```
[the remark this answers is not in the document]
Comment (mtenpow) [reply pyht/async_client.py:230]: Changed the name back to stream.

File: pyht/client.py [hunk 3]
```diff
@@ -420,3 +462,9 @@ def __next__(self) -> bytes:
 
     def close(self):
         self._close.set()
+
+
+def _audio_begins_at(fmt: Format) -> int:
+    if fmt == Format.FORMAT_WAV or fmt == Format.FORMAT_MP3:
+        return 1
+    return 0
```
[the remark this answers is not in the document]
Comment (mtenpow) [reply pyht/client.py:470]: Done

File: pyht/telemetry.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        return self.duration
+
+    def format(self):
+        return f"{self.name}:{self.duration}/{self.count}"
+
+    def __str__(self):
+        return self.format()
+
+    def __repr__(self):
+        return repr({'duration': self.duration, 'count': self.count})
```
[the remark this answers is not in the document]
Comment (mtenpow) [reply pyht/telemetry.py:165]: Fixed

File: pyht/telemetry.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        print(f"Finish {self.operation} @ {self.end_time} duration: {self.duration}")
+
+    def __repr__(self):
+        return repr(self.__dict__)
+
+
+class Timer:
+    def __init__(self, name: str, auto_finish: bool):
+        self.name = name
+        self.auto_finish = auto_finish
```
[the remark this answers is not in the document]
Comment (mtenpow) [reply pyht/telemetry.py:121]: Got rid of auto finish

File: pyht/async_client.py [hunk 6]
```diff
@@ -16,7 +16,7 @@
 from grpc.aio import Channel, Call, insecure_channel, secure_channel, UnaryStreamCall
 from grpc import ssl_channel_credentials, StatusCode
 
-
+from . import telemetry, Format
```
[the remark this answers is not in the document]
Comment (mtenpow) [reply pyht/async_client.py]: Fixed

**Review by NCarrollPlay — APPROVED** [review]

**Pull request merged**